File: src/main/scala/mffs/item/card/ItemCardLink.java
package mffs.item.card;

import resonant.api.mffs.card.ICoordLink;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import mffs.card.ItemCard;
import net.minecraft.block.Block;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.world.World;
import resonant.lib.utility.LanguageUtility;
import universalelectricity.api.vector.VectorWorld;

import java.util.List;

/**
 * A linking card used to link machines in specific positions.
 *
 * @author Calclavia
 */
public class ItemCardLink extends ItemCard implements ICoordLink
{
	public ItemCardLink(int id)
	{
		super(id, "cardLink");
	}

	@Override
	@SideOnly(Side.CLIENT)
	public void addInformation(ItemStack itemstack, EntityPlayer entityplayer, List list, boolean flag)
	{
		super.addInformation(itemstack, entityplayer, list, flag);

		if (hasLink(itemstack))
		{
			VectorWorld vec = getLink(itemstack);
			int blockId = vec.getBlockID(entityplayer.worldObj);

			if (Block.blocksList[blockId] != null)
			{
				list.add(LanguageUtility.getLocal("info.item.linkedWith") + " " + Block.blocksList[blockId].getLocalizedName());
			}

			list.add(vec.xi() + ", " + vec.yi() + ", " + vec.zi());
			list.add(LanguageUtility.getLocal("info.item.dimension") + " " + vec.world.provider.getDimensionName());
		}
		else
		{
			list.add(LanguageUtility.getLocal("info.item.notLinked"));
		}
	}

	@Override
	public boolean onItemUse(ItemStack itemStack, EntityPlayer player, World world, int x, int y, int z, int par7, float par8, float par9, float par10)
	{
		if (!world.isRemote)
		{
			VectorWorld vector = new VectorWorld(world, x, y, z);
			this.setLink(itemStack, vector);

			if (Block.blocksList[vector.getBlockID(world)] != null)
			{
				player.addChatMessage(LanguageUtility.getLocal("info.item.linkedWith") + " " + x + ", " + y + ", " + z + " - " + Block.blocksList[vector.getBlockID(world)].getLocalizedName());
			}
		}

		return true;
	}

	public boolean hasLink(ItemStack itemStack)
	{
		return getLink(itemStack) != null;
	}

	@Override
	public VectorWorld getLink(ItemStack itemStack)
	{
		if (itemStack.stackTagCompound == null || !itemStack.getTagCompound().hasKey("link"))
		{
			return null;
		}

		return new VectorWorld(itemStack.getTagCompound().getCompoundTag("link"));
	}

	@Override
	public void setLink(ItemStack itemStack, VectorWorld vec)
	{
		if (itemStack.getTagCompound() == null)
		{
			itemStack.setTagCompound(new NBTTagCompound());
		}

		itemStack.getTagCompound().setCompoundTag("link", vec.writeToNBT(new NBTTagCompound()));
	}

	public void clearLink(ItemStack itemStack)
	{
		itemStack.getTagCompound().removeTag("link");
	}
}

File: src/main/scala/mffs/item/card/ItemCardLink.scala
package mffs.item.card

import java.util
import java.util.Optional

import mffs.api.card.CoordLink
import nova.core.entity.Entity
import nova.core.game.Game
import nova.core.player.Player
import nova.core.retention.{Storable, Stored}
import nova.core.util.Direction
import nova.core.util.collection.Pair
import nova.core.util.transform.vector.{Vector3d, Vector3i}
import nova.core.world.World

/**
 * A linking card used to link machines in specific positions.
 *
 * @author Calclavia
 */
class ItemCardLink extends ItemCard with CoordLink with Storable {

	@Stored
	var linkWorld: World = null
	@Stored
	var linkPos: Vector3i = null

	override def setLink(world: World, position: Vector3i) {
		linkWorld = world
		linkPos = position
	}

	override def getLink: Pair[World, Vector3i] = new Pair(linkWorld, linkPos)

	override def getTooltips(player: Optional[Player], tooltips: util.List[String]) {
		super.getTooltips(player, tooltips)

		if (linkWorld != null && linkPos != null) {
			val block = linkWorld.getBlock(linkPos)

			if (block.isPresent) {
				tooltips.add(Game.instance.languageManager.translate("info.item.linkedWith") + " " + block.get().getID)
			}

			tooltips.add(linkPos.x + ", " + linkPos.y + ", " + linkPos.z)
			tooltips.add(Game.instance.languageManager.translate("info.item.dimension") + " " + linkWorld.getID)
		}
		else {
			tooltips.add(Game.instance.languageManager.translate("info.item.notLinked"))
		}
	}

	override def onUse(entity: Entity, world: World, position: Vector3i, side: Direction, hit: Vector3d): Boolean = {
		super.onUse(entity, world, position, side, hit)

		if (Game.instance.networkManager.isServer) {

			val block = world.getBlock(position)
			if (block.isPresent) {
				linkWorld = world
				linkPos = position
				//TODO: Fix chat msg
				//player.addChatMessage(new ChatComponentTranslation("info.item.linkedWith", x + ", " + y + ", " + z + " - " + vector.getBlock(world).getLocalizedName))
			}
		}
		return true
	}

	override def getID: String = "cardLink"
}